Disabling a disease variant for a disease leads to error if value used - Customizable Enums - Case Directory

### Bug Description

If a disease variant is being used in cases it is possible to disable the customizable enum value for the disease leading to error in the case directory.

### Reproduction Steps

1. Create a disease variant for a disease in the Customizable Enums.
2. Assign a disease variant on some cases.
3. Remove the disease variant assingment in the Customizable Enums.
4. Open Case Directory.
5. Error.


### Expected Behavior

No error should occur.

### Affected Area(s)

Web app

### SORMAS Version

1.103.0

### Android version/Browser

Any

### User Role(s)

Any

### Server URL

_No response_

### Additional Information

<img width="1167" height="591" alt="Image" src="https://github.com/user-attachments/assets/ae2988d3-7a20-457c-bdcb-b7d6a8ecf87b" />

```
sormas-payara  | Caused by: java.lang.IllegalArgumentException: Invalid enum value A+B for customizable enum type Disease variant
sormas-payara  |        at de.symeda.sormas.backend.customizableenum.CustomizableEnumFacadeEjb.getEnumValue(CustomizableEnumFacadeEjb.java:282)
sormas-payara  |        at jdk.internal.reflect.GeneratedMethodAccessor1300.invoke(Unknown Source)
sormas-payara  |        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
sormas-payara  |        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
sormas-payara  |        at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:588)
sormas-payara  |        at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:408)
sormas-payara  |        at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4835)
sormas-payara  |        at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:665)
```

--- a/sormas-api/src/main/java/de/symeda/sormas/api/i18n/Validations.java
+++ b/sormas-api/src/main/java/de/symeda/sormas/api/i18n/Validations.java
@@ -51,6 +51,7 @@ public interface Validations {
 	String contactFollowUpUntilDateSoftValidation = "contactFollowUpUntilDateSoftValidation";
 	String contactWithoutInfrastructureData = "contactWithoutInfrastructureData";
 	String customizableEnumValueAllowedCharacters = "customizableEnumValueAllowedCharacters";
+	String customizableEnumValueAlreadyInUse = "customizableEnumValueAlreadyInUse";
 	String customizableEnumValueDuplicateLanguage = "customizableEnumValueDuplicateLanguage";
 	String customizableEnumValueDuplicateValue = "customizableEnumValueDuplicateValue";
 	String customizableEnumValueEmptyTranslations = "customizableEnumValueEmptyTranslations";

--- a/sormas-api/src/main/resources/validations.properties
+++ b/sormas-api/src/main/resources/validations.properties
@@ -332,4 +332,4 @@ smsAuthKeyValueValidation = SMS Auth key value is not valid
 smsAuthSecretValueValidation = SMS Auth secret value is not valid
 systemConfigurationValueValidationInvalidBackgroundColor = Pre-defined values are: green, red, indigo, gray, default (case-sensitive) otherwise must match hexadecimal format (with or without hash "#"), example: #dd2b0e or #4AA or dd2b0e
 systemConfigurationValueValidationMenuSubtitle = Can be empty or up to 16 (any) characters. Can be used to define name of the environment: PRODUCTION - TEST etc.
-customizableEnumValueAlreadyInUse = The customizable value %s is being used for %s case(IDs) and therefore cannot be deleted or deactivated. Please remove all references to this value before attempting to deactivate.
+customizableEnumValueAlreadyInUse = The customizable value %s is being used by case IDs %s and therefore cannot be deleted or deactivated. Please remove all references to this value before attempting to deactivate.

--- a/sormas-api/src/main/resources/validations.properties
+++ b/sormas-api/src/main/resources/validations.properties
@@ -332,3 +332,4 @@ smsAuthKeyValueValidation = SMS Auth key value is not valid
 smsAuthSecretValueValidation = SMS Auth secret value is not valid
 systemConfigurationValueValidationInvalidBackgroundColor = Pre-defined values are: green, red, indigo, gray, default (case-sensitive) otherwise must match hexadecimal format (with or without hash "#"), example: #dd2b0e or #4AA or dd2b0e
 systemConfigurationValueValidationMenuSubtitle = Can be empty or up to 16 (any) characters. Can be used to define name of the environment: PRODUCTION - TEST etc.
+customizableEnumValueAlreadyInUse = The customizable value %s is using for %s case(s) and cannot be deleted or deactivated. Please update all references to this value before trying again.

--- a/sormas-api/src/main/resources/validations.properties
+++ b/sormas-api/src/main/resources/validations.properties
@@ -332,4 +332,4 @@ smsAuthKeyValueValidation = SMS Auth key value is not valid
 smsAuthSecretValueValidation = SMS Auth secret value is not valid
 systemConfigurationValueValidationInvalidBackgroundColor = Pre-defined values are: green, red, indigo, gray, default (case-sensitive) otherwise must match hexadecimal format (with or without hash "#"), example: #dd2b0e or #4AA or dd2b0e
 systemConfigurationValueValidationMenuSubtitle = Can be empty or up to 16 (any) characters. Can be used to define name of the environment: PRODUCTION - TEST etc.
-customizableEnumValueAlreadyInUse = The customizable value %s is using for %s case(s) and cannot be deleted or deactivated. Please update all references to this value before trying again.
+customizableEnumValueAlreadyInUse = The customizable value %s is being used for %s case(IDs) and therefore cannot be deleted or deactivated. Please remove all references to this value before attempting to deactivate.

--- a/sormas-backend/src/main/java/de/symeda/sormas/backend/customizableenum/CustomizableEnumFacadeEjb.java
+++ b/sormas-backend/src/main/java/de/symeda/sormas/backend/customizableenum/CustomizableEnumFacadeEjb.java
@@ -26,6 +26,7 @@ import java.util.EnumMap;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
@@ -168,7 +169,19 @@ public class CustomizableEnumFacadeEjb
 		validate(dto);
 
 		CustomizableEnumValue existingEntity = service.getByUuid(dto.getUuid());
-
+		// if existingEntity disease removed and it is mapped to the cases, shouldn't allow to save the entity
+		if (existingEntity != null && CollectionUtils.isNotEmpty(existingEntity.getDiseases()) && CollectionUtils.isNotEmpty(dto.getDiseases())) {
+			existingEntity.getDiseases().stream().filter(disease -> !dto.getDiseases().contains(disease)).findAny().ifPresent(disease -> {
+				List<String> uuids = service.areCasesUsingCustomizableEnumValue(disease, getEnumValue(dto.getDataType(), disease, dto.getValue()));
+				if (!uuids.isEmpty()) {
+					throw new ValidationRuntimeException(
+						I18nProperties.getValidationError(
+							Validations.customizableEnumValueAlreadyInUse,
+							disease.getName(),
+							uuids.stream().filter(Objects::nonNull).collect(Collectors.joining(", "))));
+				}
+			});
+		}
 		for (Disease disease : dto.getDiseases()) {
 			List<String> dataTypeValues = enumValues.get(dto.getDataType()).getOrDefault(disease, Collections.emptyList());
 			if (existingEntity == null && dataTypeValues != null && dataTypeValues.contains(dto.getValue())) {

--- a/sormas-backend/src/main/java/de/symeda/sormas/backend/customizableenum/CustomizableEnumFacadeEjb.java
+++ b/sormas-backend/src/main/java/de/symeda/sormas/backend/customizableenum/CustomizableEnumFacadeEjb.java
@@ -170,18 +170,25 @@ public class CustomizableEnumFacadeEjb
 
 		CustomizableEnumValue existingEntity = service.getByUuid(dto.getUuid());
 		// if existingEntity disease removed and it is mapped to the cases, shouldn't allow to save the entity
-		if (existingEntity != null && CollectionUtils.isNotEmpty(existingEntity.getDiseases()) && CollectionUtils.isNotEmpty(dto.getDiseases())) {
-			existingEntity.getDiseases().stream().filter(disease -> !dto.getDiseases().contains(disease)).findAny().ifPresent(disease -> {
-				List<String> uuids = service.areCasesUsingCustomizableEnumValue(disease, getEnumValue(dto.getDataType(), disease, dto.getValue()));
-				if (!uuids.isEmpty()) {
-					throw new ValidationRuntimeException(
-						I18nProperties.getValidationError(
-							Validations.customizableEnumValueAlreadyInUse,
-							disease.getName(),
-							uuids.stream().filter(Objects::nonNull).collect(Collectors.joining(", "))));
+		
+		if (existingEntity != null && CollectionUtils.isNotEmpty(existingEntity.getDiseases())) {
+			Set<Disease> incomingDiseases = dto.getDiseases() != null ? dto.getDiseases() : Collections.emptySet();
+			for (Disease removedDisease : existingEntity.getDiseases()) {
+				if (!incomingDiseases.contains(removedDisease)) {
+					List<String> uuids = service.areCasesUsingCustomizableEnumValue(
+						removedDisease,
+						getEnumValue(dto.getDataType(), removedDisease, existingEntity.getValue()));
+					if (!uuids.isEmpty()) {
+						throw new ValidationRuntimeException(
+							I18nProperties.getValidationError(
+								Validations.customizableEnumValueAlreadyInUse,
+								removedDisease.getName(),
+								uuids.stream().filter(Objects::nonNull).collect(Collectors.joining(", "))));
+					}
 				}
-			});
+			}
 		}
+
 		for (Disease disease : dto.getDiseases()) {
 			List<String> dataTypeValues = enumValues.get(dto.getDataType()).getOrDefault(disease, Collections.emptyList());
 			if (existingEntity == null && dataTypeValues != null && dataTypeValues.contains(dto.getValue())) {

--- a/sormas-backend/src/main/java/de/symeda/sormas/backend/customizableenum/CustomizableEnumValueService.java
+++ b/sormas-backend/src/main/java/de/symeda/sormas/backend/customizableenum/CustomizableEnumValueService.java
@@ -15,6 +15,10 @@
 
 package de.symeda.sormas.backend.customizableenum;
 
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.ejb.EJB;
 import javax.ejb.LocalBean;
 import javax.ejb.Stateless;
 import javax.persistence.criteria.CriteriaBuilder;
@@ -23,14 +27,22 @@ import javax.persistence.criteria.From;
 import javax.persistence.criteria.Predicate;
 import javax.persistence.criteria.Root;
 
+import de.symeda.sormas.api.Disease;
+import de.symeda.sormas.api.caze.CaseCriteria;
 import de.symeda.sormas.api.customizableenum.CustomizableEnumCriteria;
+import de.symeda.sormas.api.disease.DiseaseVariant;
 import de.symeda.sormas.api.utils.DataHelper;
+import de.symeda.sormas.backend.caze.Case;
+import de.symeda.sormas.backend.caze.CaseService;
 import de.symeda.sormas.backend.common.AdoServiceWithUserFilterAndJurisdiction;
 import de.symeda.sormas.backend.common.CriteriaBuilderHelper;
 
 @Stateless
 @LocalBean
 public class CustomizableEnumValueService extends AdoServiceWithUserFilterAndJurisdiction<CustomizableEnumValue> {
+
+	@EJB
+	private CaseService caseService;
 
 	public CustomizableEnumValueService() {
 		super(CustomizableEnumValue.class);
@@ -68,6 +80,22 @@ public class CustomizableEnumValueService extends AdoServiceWithUserFilterAndJur
 		}
 
 		return filter;
+	}
+
+	/**
+	 * Get all cases for a specific disease and disease variant that are using a customizable enum value. This is used to determine whether
+	 * a customizable enum value can be deleted or not.
+	 * 
+	 * @param disease
+	 * @param diseaseVariant
+	 * @return
+	 */
+	public List<String> areCasesUsingCustomizableEnumValue(Disease disease, DiseaseVariant diseaseVariant) {
+		CaseCriteria caseCriteria = new CaseCriteria();
+		caseCriteria.setDisease(disease);
+		caseCriteria.setDiseaseVariant(diseaseVariant);
+		List<Case> result = caseService.findBy(caseCriteria, false);
+		return result.stream().map(Case::getUuid).collect(Collectors.toList());
 	}
 
 }

--- a/sormas-backend/src/main/java/de/symeda/sormas/backend/customizableenum/CustomizableEnumValueService.java
+++ b/sormas-backend/src/main/java/de/symeda/sormas/backend/customizableenum/CustomizableEnumValueService.java
@@ -94,7 +94,7 @@ public class CustomizableEnumValueService extends AdoServiceWithUserFilterAndJur
 		CaseCriteria caseCriteria = new CaseCriteria();
 		caseCriteria.setDisease(disease);
 		caseCriteria.setDiseaseVariant(diseaseVariant);
-		List<Case> result = caseService.findBy(caseCriteria, false);
+		List<Case> result = caseService.findBy(caseCriteria, true);
 		return result.stream().map(Case::getUuid).collect(Collectors.toList());
 	}
 

